torch/_subclasses/fake_tensor.py

```diff
@@ -1400,8 +1410,20 @@ def can_fallback(func: OpOverload):
         # run kernel registered to meta for func, which include
         # python meta registrations, prims, decomps, and c++ meta fns (structured kernels)
         try:
+            common_device = FakeTensor._find_common_device(func, args, kwargs)
             with in_kernel_invocation_manager(self):
                 r = func(*args, **kwargs)
+                if (
+                    common_device is not None
+                    and common_device[0] is not None
+                    and is_noncontiguous_supported(common_device[0]) is False
+                ):
+                    from torch.distributed._tensor.ops.pointwise_ops import (
+                        pointwise_ops,
+                    )
+
+                    if func in pointwise_ops:
+                        r = r.new_empty(r.shape)
```

Comment:
This is the wrong place to put this. Modify `_make_elementwise_binary_reference` and similar functions.

Reply:
Use device_hint as seen in torch/_meta_registrations.py

Reply:
@ezyang, Thanks for the reference. 

However I still see there is a caveat going with device_hint API. When fake_device is not available, device_hint just defaults to CUDA in those cases.
We are working with 'hpu' backend and going with default API behavior can effect other's backend usecases.

At _make_elementwise_binary_reference functions fake_device is not available as its arguments get a tensor not the fakeTensor. 
Reference: _ref() function gets a tensor and doesn't have fake_device property. 
https://github.com/pytorch/pytorch/blob/main/torch/_refs/__init__.py#L939

As mentioned, This arguments modification happens in wrapper function while prepare bound.arguments and fake_device is dropped.
https://github.com/pytorch/pytorch/blob/main/torch/_prims_common/wrappers.py#L128

Reply:
I don't believe the diagnosis. In prepare bound arguments, we do dtype casting, which should also produce fake tensors that still have hpu device.

Do you have a self contained repro for the behavior you see?

Reply:
It turns out dtype casting has some issue, on casting usecase scenario it create a tensor from an input faketensor.
code: https://github.com/pytorch/pytorch/blob/main/torch/_prims_common/wrappers.py#L39

> 
```
def _maybe_convert_to_dtype(a, dtype):
    if isinstance(a, TensorLike):
        if a.dtype != dtype:
            return a.to(dtype)
        return a
```

1) Testcase to repo issue:
```
import torch

def fn(a):
 b = a.t()
 b.mul_(1.0)   # introduce dtype cast
 return b

x = torch.arange(6).reshape([2, 3])
compiled_fn_cpu = torch.compile(fn)
y_cpu = compiled_fn_cpu(x)
print("CPU:")
print(y_cpu)
```


It produces tensors as below:
```
`> /home/vrsingh/anaconda3/envs/py_venv_3.8/lib/python3.8/site-packages/torch/_prims_common/wrappers.py(51)_maybe_convert_to_dtype()
-> return a.to(dtype)
(Pdb) args
a = FakeTensor(FakeTensor(..., device='meta', size=(3, 2), dtype=torch.int64), cpu)
dtype = torch.float32
(Pdb) t_cast
tensor(..., device='meta', size=(3, 2))`
```

2) Working testcase, without dtype cast bound.arguments holds faketensors :
```
`import torch

def fn(a):
 b = a.t()
 b.mul_(1)
 return b

x = torch.arange(6).reshape([2, 3])
compiled_fn_cpu = torch.compile(fn)
y_cpu = compiled_fn_cpu(x)
print("CPU:")
print(y_cpu)` 
```

**Versions**
```
Collecting environment information...
PyTorch version: 2.0.1a0+gite034325
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 14.0.5 (ssh://gerrit:29418/tpc_llvm10 add100b010470812d36fae985e48817aaea461a2)
CMake version: version 3.26.3
Libc version: glibc-2.31

Python version: 3.8.16 (default, Mar  2 2023, 03:21:46)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-150-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 57 bits virtual
CPU(s):                          16
On-line CPU(s) list:             0-15
Thread(s) per core:              2
Core(s) per socket:              4
Socket(s):                       2
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           106
Model name:                      Intel(R) Xeon(R) Platinum 8380 CPU @ 2.30GHz
Stepping:                        6
CPU MHz:                         2294.608
BogoMIPS:                        4589.21
Virtualization:                  VT-x
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       256 KiB
L1i cache:                       256 KiB
L2 cache:                        32 MiB
L3 cache:                        32 MiB
NUMA node0 CPU(s):               0-15
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves wbnoinvd arat avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==2.0.2
[pip3] torch==2.0.1a0+gitdefeb45
[pip3] torch-tb-profiler==0.4.0
[pip3] torchmetrics==0.11.4
[pip3] torchvision==0.15.1a0+42759b1
[pip3] triton==2.0.0
[conda] Could not collect
```

Reply:
Oh, this is expected, but it also explains why you don't have exact device hints.

The reason this is expected is that the call to the decomp occurs here:

```
        # run kernel registered to meta for func, which include
        # python meta registrations, prims, decomps, and c++ meta fns (structured kernels)
        try:
            with in_kernel_invocation_manager(self):
                r = func(*args, **kwargs)
        except NotImplementedError as not_implemented_error:
            # no meta kernel registered, fallback to kernel for the device
            if has_symbolic_sizes or not can_fallback(func):
                raise UnsupportedOperatorException(func)
            return run_fallback_kernel(self, func, args, kwargs, not_implemented_error)

        return self.wrap_meta_outputs_with_default_device_logic(r, func, args, kwargs)
```

Under `in_kernel_invocation_manager`, the intended semantics is to produce meta tensors from ops, which then get manually wrapped into fake tensor at the end.

So, you CAN figure out the device hint, but you have to look at the actual inputs, not the intermediates. meta_registrations tend to have this problem less because they don't generate so many intermediate values. That's probably the best workaround for your case I guess...

Reply:
Yes, on similar lines initially i was using `_find_common_device `API to get device type from actual input args and added logic during in_kernel_invocation_manager there. 
> common_device = FakeTensor._find_common_device(func, args, kwargs)

Do you mean to go with the earlier approach? Earlier approach: https://github.com/pytorch/pytorch/pull/104689/commits/cfa2cd9b915950774156451f0ac5d40c1b0ec353
Please Confirm.

Reply:
Hi @ezyang , 
Could you please confirm if i can move to previous approach ?

 Earlier approach: https://github.com/pytorch/pytorch/commit/cfa2cd9b915950774156451f0ac5d40c1b0ec353 [ Get device type info from original tensor and add logic during in_kernel_invocation_manager]

Reply:
I'd like to split the difference. Can you move it into the dtype/device conversion in the decomp, so that we durably record the input device hint and then you can reference it inside the decomp?

Reply:
@ezyang , Thanks for the pointers.
I have added change to capture device_hint during dtype conversion in decomp. Could you please have a look at the latest change revision for feedback.

There is an open issue with the latest change which i am seeing. Getting an AssertionError in call to **from_meta_and_device** when i wrap tensor with original tensor device.
```
converter.from_meta_and_device( fake_mode, x, common_device)
```
Error:
> *** AssertionError: True, False

Any ways to resolve this error ?  Please share.